homeassistant/components/obihai/config_flow.py

```diff
@@ -41,35 +48,91 @@ async def async_validate_creds(hass: HomeAssistant, user_input: dict[str, Any])
 class ObihaiFlowHandler(ConfigFlow, domain=DOMAIN):
     """Config flow for Obihai."""
 
-    VERSION = 1
+    VERSION = 2
+    discovery_schema: vol.Schema | None = None
 
     async def async_step_user(
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
         """Handle a flow initialized by the user."""
+
         errors: dict[str, str] = {}
+        ip: str | None = None
 
         if user_input is not None:
-            self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST]})
-            if await async_validate_creds(self.hass, user_input):
-                return self.async_create_entry(
-                    title=user_input[CONF_HOST],
-                    data=user_input,
-                )
-            errors["base"] = "cannot_connect"
-
-        data_schema = self.add_suggested_values_to_schema(DATA_SCHEMA, user_input)
+            try:
+                ip = gethostbyname(user_input[CONF_HOST])
+            except gaierror:
+                errors["base"] = "cannot_connect"
+
+            if ip:
+                if pyobihai := await async_validate_creds(self.hass, user_input):
+                    device_mac = await self.hass.async_add_executor_job(
+                        pyobihai.get_device_mac
+                    )
+                    await self.async_set_unique_id(device_mac)
+                    self._abort_if_unique_id_configured()
+
+                    return self.async_create_entry(
+                        title=user_input[CONF_HOST],
+                        data=user_input,
+                    )
+                errors["base"] = "invalid_auth"
+
+        data_schema = self.discovery_schema or self.add_suggested_values_to_schema(
+            DATA_SCHEMA, user_input
+        )
         return self.async_show_form(
             step_id="user",
             errors=errors,
             data_schema=data_schema,
         )
 
+    async def async_step_dhcp(self, discovery_info: dhcp.DhcpServiceInfo) -> FlowResult:
+        """Prepare configuration for a DHCP discovered Obihai."""
+
+        return await self.async_step_dhcp_confirm(discovery_info)
+
+    async def async_step_dhcp_confirm(
+        self, discovery_info: dhcp.DhcpServiceInfo
+    ) -> FlowResult:
+        """Attempt to confirm."""
+
+        await self.async_set_unique_id(discovery_info.macaddress)
+        self._abort_if_unique_id_configured()
+
+        user_input = {
+            CONF_HOST: discovery_info.ip,
+            CONF_PASSWORD: DEFAULT_PASSWORD,
+            CONF_USERNAME: DEFAULT_USERNAME,
+        }
+
+        if await async_validate_creds(self.hass, user_input):
```

Comment:
```suggestion
        if user_input is None:
            # Show the confirmation dialog
            return self.async_show_form(step_id="dhcp_confirm")

        credentials = {
            CONF_HOST: self._dhcp_discovery_info.ip,
            CONF_PASSWORD: DEFAULT_PASSWORD,
            CONF_USERNAME: DEFAULT_USERNAME,
        }

        if await async_validate_creds(self.hass, credentials):
```

Reply:
So, I tried this locally (with a few modifications to get it working), it seems to behave the same as the previous implementation, at least from a UX perspective, so if it's preferred I have no objections to doing it this way.

I also experimented a bit with _just_ having a dialog that says "Do you want to configure this device?" in the event the credentials are default.  This works, but I'm don't really like it, because it's possible that users will want to use non-admin users, or some other customization.  IMHO it's preferrable to just let users review the configuration before submitting.  Otherwise we potentially complicate the flow.